### PR TITLE
fix: add validReactAttributes to all spread props to handle invalid DOM elements

### PR DESCRIPTION
**Pull request by dmhalejr**
`qaHook` is being added in as an attribute on the DOM elements inappropriately when components have spread operators defined on props. Let's stop that nonsense with `qahook` 🎊 

This also has a significant `<Link` attribute change on it where styles were being wiped out completely  and should be merged once ready.

### CODE REVIEW

**Comment by vacas5** [comment]
so `qahook` is not a valid dom element either, so I'm not sure this will exactly fix our issue. Can you test this branch against our other repos using `npm link` or the yarn equivalent and see if it solves the console errors we were seeing?

File: src/components/form/numberInput.jsx [hunk 1]
```diff
@@ -167,13 +168,15 @@ class NumberInput extends React.Component {
...
+
     return (
       <div
         className="NumberInput"
         style={styles.element.numberInput.container.base}
       >
         <input
-          {...props}
+          {...sanitizedProps}
           qa-hook={createQAHook(this.state.value, "number", "input")}
```
Comment (vacas5) [inline src/components/form/numberInput.jsx]: so, I think we may have another issue. Sorry I missed this on the first run. As far as I understand it `qa-hook` is not a valid HTML property. We would want `data-qa-hook` in this case, correct? https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes

File: src/components/link/index.jsx [hunk 2]
```diff
@@ -2,24 +2,29 @@ import React from "react";
...
-    </RouterLink>
-);
+const Link = (props) => {
+  const sanitizedProps = validReactAttributes(props);
+
+  return (
+    isExternal(props.to) || (!props.to && props.onClick) ?
+      <a
+        href={props.to}
+        data-qa={props.qahook ? createQAHook(`${props.qahook}`, "external-link", "link") : null}
```
Comment (vacas5) [inline src/components/link/index.jsx]: see, `data-qa` is perfectly fine. Should we change all instances on an HTML element to be `data-qa`?

File: src/components/tileVideo/index.jsx [hunk 3]
```diff
@@ -1,10 +1,11 @@
 import React from "react";
 import radium from "radium";
 import CardVideo from "../cardVideo";
+import { validReactAttributes } from "../../utils/validReactAttributes";
 
 const TileVideo = (props) => (
   <CardVideo
-    {...props}
+    {...validReactAttributes(props)}
```
Comment (vacas5) [inline src/components/tileVideo/index.jsx]: this is a React component. We wouldn't need to sanitize here.

File: src/components/textarea/index.jsx [hunk 4]
```diff
@@ -107,9 +108,11 @@ class Textarea extends React.Component {
     delete props.maxLines;
     delete props.disableEnter;
```
Comment (vacas5) [inline src/components/textarea/index.jsx]: guess what all these lines were trying to do

File: src/components/text/textSuper.jsx [hunk 5]
```diff
@@ -4,10 +4,11 @@ import radium from "radium";
 import Heading from "./textHeading";
 import { textSuper } from "../../utils/typography";
 import propTypes from "../../utils/propTypes";
+import { validReactAttributes } from "../../utils/validReactAttributes";
 
 const TextSuper = (props) => (
   <Heading
-    {...props}
+    {...validReactAttributes(props)}
```
Comment (vacas5) [inline src/components/text/textSuper.jsx]: `Heading` is a React component. Don't want to sanitize here

File: src/components/link/index.jsx (hunk 2 above)
Comment (dmhalejr) [reply src/components/link/index.jsx]: Correct. This was an incorrect placement of mine in two components. Moving them to be `data-qa` properly. Thanks 🦅 👁

File: src/components/form/numberInput.jsx (hunk 1 above)
Comment (dmhalejr) [reply src/components/form/numberInput.jsx]: Fixed!

File: src/components/tabs/tab.jsx [hunk 6]
```diff
@@ -2,9 +2,10 @@ import React from "react";
 import PropTypes from "prop-types";
 import radium from "radium";
 import { NavigationTab } from "../navigation";
+import { validReactAttributes } from "../../utils/validReactAttributes";
 
 const Tab = (props) => (
-  <NavigationTab {...props}>
+  <NavigationTab {...validReactAttributes(props)}>
```
Comment (vacas5) [inline src/components/tabs/tab.jsx]: no need to sanitize here, as this is a React component

File: src/components/settingBlockTextArea/index.jsx [hunk 7]
```diff
@@ -17,7 +18,7 @@ const SettingBlockTextArea = (props) => {
       <HeightExpander idToFind={props.id} baseHeight="0px">
         {(expandHeight, newHeight) => (
           <TextArea
-            {...props}
+            {...validReactAttributes(props)}
```
Comment (vacas5) [inline src/components/settingBlockTextArea/index.jsx]: react component, don't sanitize

File: src/components/settingBlockInput/index.jsx [hunk 8]
```diff
@@ -12,7 +13,7 @@ const SettingBlockInput = (props) => (
     htmlFor={props.id}
   >
     <Input
-      {...props}
+      {...validReactAttributes(props)}
```
Comment (vacas5) [inline src/components/settingBlockInput/index.jsx]: React component, don't sanitize

File: src/components/loading/index.jsx [hunk 9]
```diff
@@ -22,12 +23,16 @@ const styles = {
...
-    {...props}
-    style={Object.assign({}, styles, props.style)}
-  />
-);
+const Loading = (props) => {
+  const sanitizedProps = validReactAttributes(props);
+
+  return (
+    <Icon
+      {...sanitizedProps}
```
Comment (vacas5) [inline src/components/loading/index.jsx]: react component, don't sanitize

File: src/components/icon/index.jsx [hunk 10]
```diff
@@ -83,222 +84,222 @@ Icon.defaultProps = {
 // Interface
 
 exports.ArrowLeft = props => (
-  <Icon {...props}>
+  <Icon {...validReactAttributes(props)}>
```
Comment (vacas5) [inline src/components/icon/index.jsx]: none of these changes should be introduced. `Icon` is a React component. Fix the rest spread in the `Icon` component itself

File: src/components/textarea/index.jsx (hunk 4 above)
Comment (dmhalejr) [reply src/components/textarea/index.jsx]: Oh man 🤦‍♂️ removed.

File: src/components/tileVideo/index.jsx (hunk 3 above)
Comment (dmhalejr) [reply src/components/tileVideo/index.jsx]: this has been... sanitized 😎 (_The Who starts playing_)

File: src/components/text/textSuper.jsx (hunk 5 above)
Comment (dmhalejr) [reply src/components/text/textSuper.jsx]: this has been... sanitized 😎 (_The Who starts playing_)

File: src/components/tabs/tab.jsx (hunk 6 above)
Comment (dmhalejr) [reply src/components/tabs/tab.jsx]: this has been... sanitized 😎 (_The Who starts playing_)

File: src/components/settingBlockTextArea/index.jsx (hunk 7 above)
Comment (dmhalejr) [reply src/components/settingBlockTextArea/index.jsx]: this has been... sanitized 😎 (_The Who starts playing_)

File: src/components/settingBlockInput/index.jsx (hunk 8 above)
Comment (dmhalejr) [reply src/components/settingBlockInput/index.jsx]: this has been... sanitized 😎 (_The Who starts playing_)

File: src/components/loading/index.jsx (hunk 9 above)
Comment (dmhalejr) [reply src/components/loading/index.jsx]: this has been... sanitized 😎 (_The Who starts playing_)

File: src/components/link/index.jsx [hunk 11]
```diff
@@ -2,24 +2,29 @@ import React from "react";
...
-      className={className}
-    >
-      {children}
-    </a>
-    :
-    <RouterLink to={to} onClick={onClick} data-qa={qaHook}>
-      {children}
-    </RouterLink>
-);
+const Link = (props) => {
```
Comment (vacas5) [inline src/components/link/index.jsx]: in an effort to change less and avoid regression, I would suggest `const Link = ({ to, onClick, children, qaHook, className, ...rest }) => (`

File: src/components/link/index.jsx [hunk 12]
```diff
@@ -2,24 +2,29 @@ import React from "react";
...
+  const sanitizedProps = validReactAttributes(props);
```
Comment (vacas5) [inline src/components/link/index.jsx]: and then on this line `const sanitizedProps = validReactAttributes(rest);`

File: src/components/link/index.jsx [hunk 13]
```diff
@@ -2,24 +2,29 @@ import React from "react";
...
+        href={props.to}
+        data-qa={props.qaHook ? createQAHook(`${props.qaHook}`, "external-link", "link") : null}
+        onClick={props.onClick}
+        {...sanitizedProps}
+      >
+        {props.children}
+      </a>
+      :
+      <RouterLink to={props.to} onClick={props.onClick}>
+        {props.children}
```
Comment (vacas5) [inline src/components/link/index.jsx]: by using `...rest` above you won't have to change all of these

File: src/components/link/index.jsx [hunk 14]
```diff
@@ -2,24 +2,29 @@ import React from "react";
...
 
 const isExternal = (url) => /^(http|https):\/\//.test(url || "");
 
-const Link = ({ to, onClick, children, qaHook, className }) => (
-  isExternal(to) || (!to && onClick) ?
-    <a
-      href={to}
-      data-qa={qaHook ? createQAHook(`${qaHook}`, "external-link", "link") : null}
-      onClick={onClick}
-      className={className}
```
Comment (vacas5) [inline src/components/link/index.jsx]: I know it's technically getting spread but maybe we just leave `className` to change fewer lines?

File: src/components/loading/index.jsx (hunk 9 above)
Comment (vacas5) [reply src/components/loading/index.jsx]: I don't think we need to sanitize here. `Icon` is a react component

Comment (dmhalejr) [reply src/components/loading/index.jsx]: 🤕

File: src/components/link/index.jsx (hunk 13 above)
Comment (dmhalejr) [reply src/components/link/index.jsx]: Confused by this statement.

File: src/components/link/index.jsx (hunk 14 above)
Comment (dmhalejr) [reply src/components/link/index.jsx]: Agreed.

File: src/components/link/index.jsx (hunk 12 above)
Comment (dmhalejr) [reply src/components/link/index.jsx]: I think I did this properly 🤞

File: src/components/link/index.jsx (hunk 13 above)
Comment (dmhalejr) [reply src/components/link/index.jsx]: I think I got it taken care of but take a look.

**Pull request merged**